### ISSUE TITLE
bug in #turn method.

### ISSUE THREAD

**Issue by azemoh** [learn-co-students/ttt-8-turn-q-000#258]
In the solution #turn executes #display_board the number of times the player enters an invalid move. so if a player makes 5 wrong moves, #turn executes #display_board 5 times when the player eventually makes a valid move.

**Comment by thegands** [comment]
i got curious about this issue and tested it myself by copying the #turns method from the solution branch. you're totally right, the board is printed out the number of times invalid input is entered (plus one for the final valid input). going over the code, i cannot see how this would be occur because turn(board) should restart the #turn method and prevent display_board(board) from being interpreted.. i think? now that i'm thoroughly confused, an explanation would be really helpful :)

**Comment by azemoh** [comment]
Yeah looking at the code you would expect to break out of the #turn method by calling it recursively, but that is not the case. I submitted a pull request #259 to fix this where i used __return__ to break out of the method.
```ruby
else
  turn(board)
  return
end
```

**Comment by aviflombaum** [comment]
@thegands @azemoh @fossbender

You all are correct. The way the `#turn` solution is implemented using recurssion would queue up extra calls to `display_board` for each invalid input. Here's what's happening

1. turn is called - let's call this turn_a
1a. within turn_a we ask for user input.
1b. turn_a finds that input invalid and during it's execution, starts another execution of turn, called turn_b
2. turn is called - let's call this turn_b
2a. within turn_b we ask for user input.
2b. turn_b finds that input invalid and during its execution, starts another execution of turn, called turn_c.
etc....

once there is valid input, each the calls to turn must resolve.

2c. calls display_board within turn_b, thereby resolving the execution of the method, so now turn_a can begin resolving.
1c. turn_a calls display_board 

I think the `return` solution that @azemoh makes sense though I'd write it like this:

```ruby
def turn(board)
  puts "Please enter 1-9:"
  input = gets.strip
  if valid_move?(board, input)
    move(board, input)
  else
    turn(board) and return
  end
  display_board(board)
end
```

**Comment by aviflombaum** [comment]
Fixed in a9ec529

**Comment by azemoh** [comment]
@aviflombaum sorry for bringing this up again but I got this issue while working on [OO TicTacToe](https://github.com/learn-co-students/oo-tic-tac-toe-q-000). I tried to use `turn() and return` and still got extra calls to #display_board. Only `turn() return` seem to work as expected. I am going to look up the `and` keyword.

**Comment by aviflombaum** [comment]
Fascinating!!! I'm going to look into it too. Thanks for bringing this up - let's also chat - ping me at avi@learn.co or find me on slack https://slack.learn.co

I sincerely hope you're planning on joining Learn as a student. You're gonna crush it.

**Comment by asherfsullivan** [comment]
this can be fixed by calling the `display_board(board)` method in side the `if` block as follows:

```
def turn(board)
  puts "Please enter 1-9:"
  input = gets.strip
  if valid_move?(board, input)
    move(board, input)
    display_board(board)
  else
    turn(board)
  end
end
```

**Comment by azemoh** [comment]
Thumbs up @asherfsullivan! this is a better implementation.
